### PR TITLE
OCM-24713 | feat(machine-pool): add node_drain_grace_period attribute

### PR DESCRIPTION
**Pull request by amandahla**
## PR Summary
  Adds optional `node_drain_grace_period` field to `aws_node_pool` in the machine-pool module and raises the minimum rhcs provider constraint to >= 1.7.7 where the attribute is first supported.

  ## Detailed Description of the Issue
  `rhcs_hcp_machine_pool` exposes `node_drain_grace_period` (minutes) since provider v1.7.7, but the machine-pool module's `aws_node_pool` variable did not surface it, leaving consumers unable to control how long the provider waits before forcibly terminating draining nodes. This PR
  wires the field through, validates the accepted range (0–10080), and bumps both the module and root provider minimum constraints from >= 1.7.3/1.7.6 to >= 1.7.7.

  Also included: tooling fixes to `Makefile` and `scripts/verify-gen.sh` that make `make verify-gen` and `make lint` more robust in isolated CI environments (Prow), replacing git-status-based doc drift detection with SHA-based hashing and isolating `terraform init` to a temp
  directory.

  ## Related Issues and PRs
  - Jira: [OCM-24713](https://issues.redhat.com/browse/OCM-24713)
  - Fixes: `#`
  - Related PR(s):
  - Related design/docs:

  ## Type of Change
  
  - [x] feat - adds a new module capability or new user-facing behavior.
  - [ ] fix - resolves incorrect module behavior or bug.
  - [ ] docs - updates documentation only.
  - [ ] style - formatting/naming changes with no logic impact.
  - [ ] refactor - module code restructuring with no behavior change.
  - [ ] test - adds or updates tests only.
  - [ ] chore - maintenance work (tooling, housekeeping, non-product code).
  - [ ] build - changes build system, packaging, or dependencies for build output.
  - [ ] ci - changes CI pipelines, jobs, or automation workflows.
  - [ ] perf - improves performance without changing intended behavior.

  ## Previous Behavior
  `aws_node_pool` had no `node_drain_grace_period` field. The rhcs provider used its own default drain timeout; module consumers had no way to override it.

  ## Behavior After This Change
  `aws_node_pool.node_drain_grace_period` (optional number) accepts 0–10080 minutes and is passed directly to `rhcs_hcp_machine_pool`. Omitting it preserves the provider default. Values outside [0, 10080] fail `terraform plan` with a validation error.

  ## How to Test (Step-by-Step)

  ### Preconditions
  - Terraform >= 1.5.7
  - rhcs provider >= 1.7.7
  - `RHCS_TOKEN` and AWS credentials set

  ### Test Steps
  1. `make pre-push-checks`
  2. `cd modules/machine-pool && terraform test` (runs `aws_node_pool.tftest.hcl` including the three new `node_drain_grace_period` run blocks)
  3. In a live environment: deploy a machine pool with `node_drain_grace_period = 60` and confirm the value appears in `terraform show` output.

  ### Expected Results
  - All unit tests pass, including `invalid_node_drain_grace_period_fails`, `valid_node_drain_grace_period_plan`, and `node_drain_grace_period_null_plan`.
  - Setting `node_drain_grace_period = 10081` fails plan validation.
  - Live deployment reflects the configured drain timeout.

  ## Proof of the Fix
  - Screenshots:
  - Videos:
  - Logs/CLI output:
  - Other artifacts:

  ## Breaking Changes
  - [x] No breaking changes
  - [ ] Yes, this PR introduces a breaking change (describe impact and migration plan below)

  ### Breaking Change Details / Migration Plan
  The minimum rhcs provider constraint is raised from >= 1.7.3 (module) / >= 1.7.6 (root) to >= 1.7.7. Consumers pinned below 1.7.7 must upgrade their provider before applying. The `aws_node_pool` interface change is additive (new optional field with null default).

  ## Developer Verification Checklist
  - [ ] **AWS-only changes:** If this PR is mainly **AWS-only** (no `rhcs` resources/variables), I linked **official Red Hat or cited ROSA HCP documentation** that supports reference alignment, or I explained why the change still belongs in-repo per **`Module scope (AWS-only vs core
  HCP)`** in [`.cursor/rules/rosa-hcp-terraform.mdc`](.cursor/rules/rosa-hcp-terraform.mdc).
  - [ ] I checked if this affects terraform-rhcs-rosa-classic and submitted (or already submitted) a companion PR when needed.
  - [x] Commit subject/title follows `[JIRA-TICKET] | [TYPE]: <MESSAGE>`.
  - [x] PR description clearly explains both **what** changed and **why**.
  - [x] Relevant Jira/GitHub issues and related PRs are linked.
  - [x] Tests were added/updated where appropriate.
  - [ ] I manually tested the change.
  - [ ] `make pre-push-checks` passes (or each step: `verify`, `verify-gen`, `lint`, `unit-tests`, `license-check`, `docs-lint`).
  - [x] Documentation was added/updated where appropriate (see `make terraform-docs`).
  - [x] Any risk, limitation, or follow-up work is documented.

## Release Notes

* **New Features**
  * Added support for configuring node drain grace period in machine pools.
  * Added capacity reservation preference option for AWS node pools.

* **Tests**
  * Added test coverage for node drain grace period behavior.

* **Chores**
  * Updated minimum rhcs provider version to 1.7.7.
  * Reordered pre-push validation checks.

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -391,6 +391,7 @@ variable "machine_pools" {
       additional_security_group_ids   = optional(list(string))
       capacity_reservation_id         = optional(string)
       capacity_reservation_preference = optional(string)
+      node_drain_grace_period         = optional(number)
```
[the remark this answers is not in the document]
Comment (amandahla) [reply variables.tf:400]: Updated to 1.7.7

Comment (olucasfreitas) [inline variables.tf:400]: The linked Jira asks for nullable-safe validation on both the submodule and the root `machine_pools` map, but the root variable currently only widens the type. Can we mirror the range check here and add a small root-path test so invalid `machine_pools[*].aws_node_pool.node_drain_grace_period` values fail at the public API boundary too?

Comment (olucasfreitas) [reply variables.tf:400]: I still think this one matters. The linked Jira explicitly asks for nullable-safe validation at both the submodule and the root `machine_pools` map, and the current diff only adds the submodule-side check.

File: versions.tf [hunk 2]
```diff
@@ -10,7 +10,7 @@ terraform {
       version = ">= 6.44.0"
     }
     rhcs = {
-      version = ">= 1.7.6"
+      version = ">= 1.7.7"
```
Comment (olucasfreitas) [inline versions.tf:13]: The provider floor is now `>= 1.7.7`, but the generated requirement tables are still stale (`README.md` says `>= 1.7.6`, `modules/machine-pool/README.md` says `>= 1.7.3`). `make verify-gen` rewrites both files locally, so please commit the regenerated docs in this PR.

File: modules/machine-pool/tests/aws_node_pool.tftest.hcl [hunk 3]
```diff
@@ -153,3 +153,83 @@ run "apply_with_capres_preference" {
...
+
+    aws_node_pool = {
+      instance_type           = "m5.xlarge"
+      tags                    = {}
+      node_drain_grace_period = 60
+    }
+  }
+
+  assert {
+    condition     = var.aws_node_pool.node_drain_grace_period == 60
```
Comment (olucasfreitas) [inline modules/machine-pool/tests/aws_node_pool.tftest.hcl]: The new happy-path test only checks the input variable, so it doesn’t really guard the `aws_node_pool -> rhcs_hcp_machine_pool` wiring. A small mocked-plan assertion on `rhcs_hcp_machine_pool.machine_pool.aws_node_pool.node_drain_grace_period` would make this coverage more meaningful.

File: variables.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (olucasfreitas) [reply variables.tf:400]: This one looks resolved now that both `versions.tf` files require `terraform-redhat/rhcs >= 1.7.7`, which is the first tagged provider release that includes `aws_node_pool.node_drain_grace_period` in `rhcs_hcp_machine_pool`.

Comment (amandahla) [reply variables.tf:400]: Good point, fixed.

File: versions.tf (hunk 2 above)
Comment (amandahla) [reply versions.tf:13]: Sure, docs updated.

File: modules/machine-pool/tests/aws_node_pool.tftest.hcl (hunk 3 above)
Comment (amandahla) [reply modules/machine-pool/tests/aws_node_pool.tftest.hcl]: Test updated too validating value from a mocked provider.

**Comment by amandahla** [comment]
@coderabbitai review the changes

**Comment by amandahla** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> 🧩 Analysis chain
> 🏁 Script executed:
> 
> ```shell
> #!/bin/bash
> echo "=== Root variables.tf ===" 
> cat variables.tf
> 
> echo ""
> echo "=== modules/machine-pool/variables.tf ==="
> cat modules/machine-pool/variables.tf
> 
> echo ""
> echo "=== modules/machine-pool/main.tf ==="
> cat modules/machine-pool/main.tf
> 
> echo ""
> echo "=== modules/machine-pool/tests/aws_node_pool.tftest.hcl ==="
> cat modules/machine-pool/tests/aws_node_pool.tftest.hcl
> 
> echo ""
> echo "=== modules/machine-pool/README.md (node_drain section) ==="
> grep -n "node_drain" modules/machine-pool/README.md
> 
> echo ""
> echo "=== README.md (node_drain section) ==="
> grep -n "node_drain" README.md
> ```
> 
> Length of output: 36546
> 
> `@amandahla`: Here is my review of the PR changes.
> 
> ## Review: [OCM-24713](https://redhat.atlassian.net/browse/OCM-24713) — `node_drain_grace_period` attribute
> ### Overall
> The implementation is clean and well-structured. The new attribute is correctly declared, validated, and threaded through (via the whole-object pass-through `aws_node_pool = var.aws_node_pool` in `main.tf`). The three new test cases cover the key scenarios. A few minor observations below.
> 
> ### `modules/machine-pool/variables.tf` & `variables.tf`
> **1. `optional(number)` accepts fractional values**
> 
> Terraform's `number` type accepts floats (e.g., `60.5`). The attribute semantics are "integer minutes," but there is no validation guard against a non-integer value like `1440.5`. If the RHCS provider's API enforces integer-only values, Terraform would accept the input and then fail at apply time with a provider-level error instead of a clear validation message.
> 
> Consider adding a floor/ceiling check:
> 
> ```hcl
> validation {
>   condition = var.aws_node_pool.node_drain_grace_period == null ? true : (
>     var.aws_node_pool.node_drain_grace_period >= 0 &&
>     var.aws_node_pool.node_drain_grace_period <= 10080 &&
>     floor(var.aws_node_pool.node_drain_grace_period) == var.aws_node_pool.node_drain_grace_period
>   )
>   error_message = "node_drain_grace_period must be a whole number between 0 and 10080 minutes (7 days)."
> }
> ```
> 
> The same applies to the root `variables.tf` validation block.
> 
> ### `modules/machine-pool/tests/aws_node_pool.tftest.hcl`
> **2. No boundary-value tests (0 and 10080)**
> 
> The `valid_node_drain_grace_period_plan` run tests `60`, which is well within range. It would be good practice to also test the inclusive boundaries `0` (minimum) and `10080` (maximum) to confirm the `>=`/`<=` operators are correctly anchored. A single extra run with `0` is the most critical edge case (a "disable drain" semantic).
> 
> **3. No test for a negative value**
> 
> A test with e.g. `node_drain_grace_period = -1` with `expect_failures = [var.aws_node_pool]` would explicitly document that negative inputs are rejected. This is especially useful given that `0` carries valid meaning (immediate eviction).
> 
> **4. Inconsistent provider alias between `valid_node_drain_grace_period_plan` and `node_drain_grace_period_null_plan`**
> 
> `valid_node_drain_grace_period_plan` uses `rhcs.no_override` while `node_drain_grace_period_null_plan` uses `rhcs.with_override`. The reason is sensible (a null computed field needs the mock default to be assertable), but a short inline comment would make the intent clear for future contributors:
> 
> ```hcl
> # Uses with_override so the computed aws_node_pool object is fully known at plan time,
> # allowing the null assertion to be evaluated.
> run "node_drain_grace_period_null_plan" {
> ```
> 
> ### `modules/machine-pool/main.tf`
> No changes needed — `node_drain_grace_period` flows automatically because `aws_node_pool` is passed as a whole object.
> 
> ### Documentation
> Both `README.md` and `modules/machine-pool/README.md` are correctly regenerated and show the new `node_drain_grace_period = optional(number)` field.
> 
> ### Summary
> Severity	Item
> ⚠️ Minor	`number` type allows floats; consider adding an integer guard in both validation blocks
To be addressed by the provider.
> 💡 Suggestion	Add boundary tests for `0` and `10080`
Nitpicking
> 💡 Suggestion	Add a negative-value rejection test
Nitpicking
> 💡 Suggestion	Add a comment explaining why `with_override` is used for the null test
Nitpicking

File: tests/machine_pools.tftest.hcl [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        aws_node_pool = {
+          instance_type           = "m5.xlarge"
+          tags                    = {}
+          node_drain_grace_period = 10081
+        }
+      }
+    }
+  }
+
+  expect_failures = [
```
Comment (olucasfreitas) [inline tests/machine_pools.tftest.hcl]: I ran `terraform test` at the repo root and this fixture fails before it reaches `expect_failures = [var.machine_pools]`: unrelated `oidc_config_id`, AWS account ID, and AZ validations fire first. Can we either seed the minimum valid root inputs/data-source mocks here or move this into a smaller fixture so it actually tests the root `machine_pools` boundary?

**Comment by olucasfreitas** [comment]
The provider already models this as `Int64`, but the module still accepts `60.5` and only rejects it downstream in provider validation. Since Jira and the provider docs describe whole minutes, I’d still consider a simple integer guard worthwhile if we want the module’s fail-fast behavior to match the public contract.

File: tests/machine_pools.tftest.hcl (hunk 4 above)
Comment (amandahla) [reply tests/machine_pools.tftest.hcl]: Good catch. After thinking it through, I've decided to drop both the root validation and the test rather than seed more fixture inputs.
  
  The duplication creates a maintenance risk: if the constraint ever changes (e.g. the max minutes gets bumped), it has to be updated in both variables.tf and modules/machine-pool/variables.tf. Since node_drain_grace_period is a purely per-pool constraint with no cross-field
  dependency at the root level, the submodule is the right place to own it — which is consistent with how the rest of the fields in var.machine_pools are handled (no root-level validations for submodule-owned field constraints).

  The coverage is preserved in modules/machine-pool/tests/aws_node_pool.tftest.hcl via invalid_node_drain_grace_period_fails

**Comment by amandahla** [comment]
> The provider already models this as `Int64`, but the module still accepts `60.5` and only rejects it downstream in provider validation. Since Jira and the provider docs describe whole minutes, I’d still consider a simple integer guard worthwhile if we want the module’s fail-fast behavior to match the public contract.

Good catch, I added one for integer and positive number

File: variables.tf (hunk 1 above)
Comment (olucasfreitas) [reply variables.tf:400]: The latest revision drops the root `machine_pools` validation again, which puts the implementation back out of sync with `ROSAENG-6788`. Can we explicitly narrow the Jira/PR scope to module-only validation?

**Comment by amandahla** [comment]
/override ci/prow/rosa-hcp-private
/override ci/prow/rosa-hcp-public

Failed due error " The vpc 'vpc-02763f495de8418b5' has dependencies and cannot be deleted.", not related to change.

File: variables.tf (hunk 1 above)
Comment (amandahla) [reply variables.tf:400]: Discussed in further comments, decided to not add validation to root. Thanks!
